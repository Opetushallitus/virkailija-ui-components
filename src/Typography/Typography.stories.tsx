import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { select } from '@storybook/addon-knobs';

import Typography from './index';

const stories = storiesOf('Components/Typography', module);

stories.add('Default', () => {
  const variant = select(
    'Variant',
    {
      Body: 'body',
      Secondary: 'secondary',
      H1: 'h1',
      H2: 'h2',
      H3: 'h3',
      H4: 'h4',
      H5: 'h5',
      H6: 'h6',
    },
    'body',
  ) as any;

  return (
    <Typography variant={variant}>
      The quick brown fox jumps over the lazy dog
    </Typography>
  );
});
