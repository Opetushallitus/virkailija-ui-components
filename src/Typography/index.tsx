import * as React from 'react';
import styled from 'styled-components';

import {
  space,
  SpaceProps,
  color,
  ColorProps,
  typography,
  TypographyProps as SystemTypographyProps,
} from '../system';

type SystemProps = SpaceProps & ColorProps & SystemTypographyProps;

type TypographyVariant =
  | 'body'
  | 'secondary'
  | 'h1'
  | 'h2'
  | 'h3'
  | 'h4'
  | 'h5'
  | 'h6';

const TypographyBase = styled.span<
  { variant: TypographyVariant } & SystemProps
>`
  margin: 0px;
  padding: 0px;
  ${({ theme, variant }) => theme.typography[variant]}
  ${space}
  ${color}
  ${typography}
`;

type TypographyBaseProps = {
  variant?: TypographyVariant;
  children: React.ReactNode;
} & SystemProps;

export type TypographyProps = TypographyBaseProps &
  Omit<React.ComponentProps<typeof TypographyBase>, keyof TypographyBaseProps>;

const variantAsMap = {
  body: 'span',
  secondary: 'span',
  h1: 'h1',
  h2: 'h2',
  h3: 'h3',
  h4: 'h4',
  h5: 'h5',
  h6: 'h6',
} as const;

const Typography = ({
  as: asProp,
  variant = 'body',
  children,
  ...props
}: TypographyProps) => {
  const as = asProp || variantAsMap[variant] || 'span';

  return (
    <TypographyBase as={as} variant={variant} {...(props as any)}>
      {children}
    </TypographyBase>
  );
};

export default Typography;
