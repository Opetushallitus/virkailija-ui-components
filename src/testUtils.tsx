import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { createTheme } from './createTheme';

const theme = createTheme();

const Providers = ({ children }: { children: React.ReactNode }) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
);

const dummyDiv = document.createElement('div');
dummyDiv.setAttribute('data-nothing', 'true');

// Return rootElement from render() for easier snapshot testing
const customRender = (ui: React.ReactElement, options: any = {}) => {
  const result = render(ui, { wrapper: Providers, ...options });
  return {
    ...result,
    rootElement: result.container.firstChild || dummyDiv,
  };
};

export * from '@testing-library/react';

export { customRender as render };
