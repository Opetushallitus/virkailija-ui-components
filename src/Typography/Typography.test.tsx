import * as React from 'react';
import { render } from '../testUtils';
import Typography from './index';

test('renders correctly', () => {
  const { rootElement } = render(<Typography>Hello world!</Typography>);
  expect(rootElement).toMatchSnapshot();
});

test('renders correctly with variant', () => {
  const { rootElement } = render(
    <>
      <Typography variant="body">Body</Typography>
      <Typography variant="h1">Heading 1</Typography>
      <Typography variant="h2">Heading 2</Typography>
      <Typography variant="h3">Heading 3</Typography>
      <Typography variant="h4">Heading 4</Typography>
      <Typography variant="h5">Heading 5</Typography>
      <Typography variant="h6">Heading 6</Typography>
    </>,
  );

  expect(rootElement).toMatchSnapshot();
});
