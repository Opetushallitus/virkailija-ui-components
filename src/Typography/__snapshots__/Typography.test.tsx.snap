// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
.c0 {
  margin: 0px;
  padding: 0px;
  line-height: 1.5;
  font-family: 'Roboto',sans-serif;
  font-size: 1rem;
  color: #666;
}

<span
  class="c0"
>
  Hello world!
</span>
`;

exports[`renders correctly with variant 1`] = `
.c0 {
  margin: 0px;
  padding: 0px;
  line-height: 1.5;
  font-family: 'Roboto',sans-serif;
  font-size: 1rem;
  color: #666;
}

<span
  class="c0"
>
  Body
</span>
`;
